### PR TITLE
Simplify the two-level iterator

### PR DESCRIPTION
**Pull request by vmx**
This commit makes the code to skip empty data blocks a bit clearer.

I have to admit that I don't understand the iterator code 100%, but I think my change makes sense and also `make check` passes locally.

### CODE REVIEW

File: table/two_level_iterator.cc [hunk 1]
```diff
@@ -154,14 +154,8 @@ void TwoLevelIterator::Prev() {
 
 
 void TwoLevelIterator::SkipEmptyDataBlocksForward() {
-  while (second_level_iter_.iter() == nullptr ||
-         (!second_level_iter_.Valid() &&
-         !second_level_iter_.status().IsIncomplete())) {
-    // Move to next block
-    if (!first_level_iter_.Valid()) {
```
Comment (igorcanadi) [inline table/two_level_iterator.cc]: Can you explain why you can assume that `first_level_iter_.Valid()`?

Preferably in the comments in the code.

File: table/two_level_iterator.cc [hunk 2]
```diff
@@ -154,14 +154,8 @@ void TwoLevelIterator::Prev() {
 
 
 void TwoLevelIterator::SkipEmptyDataBlocksForward() {
```
Comment (igorcanadi) [inline table/two_level_iterator.cc]: You added an explicit requirement for this function that `second_level_iter != nullptr`. It would be good to document this, preferably with `assert(second_level_iter.iter() != nullptr)`

**Comment by facebook-github-bot-6** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

File: table/two_level_iterator.cc (hunk 1 above)
Comment (vmx) [inline table/two_level_iterator.cc]: From my debugging experience the `first_level_iter_` was only invalid when `second_level_iter_.iter() == nullptr`. Hence I thought about moving this case out/removing that case.

File: table/two_level_iterator.cc (hunk 2 above)
Comment (vmx) [inline table/two_level_iterator.cc]: Yes indeed. It was (from my debugging experience again) always the case. Adding an assert and re-running the tests is a good idea.

**Comment by vmx** [comment]
What are the next steps?

File: table/two_level_iterator.cc [hunk 3]
```diff
@@ -158,14 +158,9 @@ void TwoLevelIterator::Prev() {
...
-  while (second_level_iter_.iter() == nullptr ||
```
Comment (IslamAbdelRahman) [inline table/two_level_iterator.cc]: We changed the while to and if condition, looks to me that this is based on the assumption that we dont have empty data blocks, I cannot think of a situation right now where we will have empty data blocks but I dont think that this is a safe assumption. It maybe safe to have this assumption with block based table but I am not sure about other table formats (plain, cuckoo) 

By empty data blocks I mean that after creating the second_level_iter_ for the new block we have second_level_iter_.Valid() == false

Comment (vmx) [inline table/two_level_iterator.cc]: My original intention was to make things clearer. But it looks like there are not much assumptions that can be made. I find the resulting code hard to follow and understand. For me it's not clear if all conditions within the loop can happen independently of the parts of the while condition of if they are somehow connected. Though it sounds like they are independent, so things need to stay like this.

**Comment by vmx** [comment]
I'm closing this issue. This is really core functionality and although I find it sad, I can see the point of not changing it (even if tests pass).
